Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp

```diff
@@ -1051,6 +1063,17 @@ void CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly()
 #endif
                 if (localDelta.eventRegionChanged)
                     state.eventRegion = eventRegion();
+                if (localDelta.damageChanged) {
+                    state.damage = Damage();
+                    if (m_needsDisplay.completeLayer)
+                        state.damage.add(FloatRect({ }, m_size));
+                    else {
+                        for (const auto& rect : m_needsDisplay.rects)
+                            state.damage.add(rect);
+                    }
+                    if (m_damagedRectsAreUnreliable)
+                        state.damage.invalidate();
```

Comment:
if `m_damagedRectsAreUnreliable` is true we set the damage to invalidate it later. If we handle invalidation as damage full layer, maybe we can do:

```
if (m_needsDisplay.completeLayer || m_damagedRectsAreUnreliable)
```

Reply:
Read my comment above about this: if the layer has 3D transformations it may cover a different part of the frame than the 2D rectangle that matches its size. OTOH, your suggestion made me realize that we can shortcut some work this way:

```suggestion
                    if (m_damagedRectsAreUnreliable)
                        state.damage.invalidate();
                    else if (m_needsDisplay.completeLayer)
                        state.damage.add(FloatRect({ }, m_size));
                    else {
                        for (const auto& rect : m_needsDisplay.rects)
                            state.damage.add(rect);
                    }
```

(That is: if we need to consider the whole frame damaged, just invalidate the `Damage` instance instead of first adding rects to it and then invalidate it as well.)